java-client/weather-server-example/src/main/docker/app.d/weather_server.py

```diff
@@ -0,0 +1,194 @@
+from dataclasses import dataclass
+from datetime import datetime
+from deephaven import Aggregation as agg
+from deephaven import DateTimeUtils as TimeUtils
+from deephaven import DynamicTableWriter, Types as dht
+from deephaven import as_list
+from deephaven.TableTools import *
+from math import cos, asin, sqrt, pi
+# import time
+from threading import Lock
+from threading import Thread
+
+
+@dataclass
+class Location:
+    city: str = None
+    state: str = None
+    lat: float = -999
+    lon: float = -999
+
+    lastObsTime: datetime = None
+
+    observationStation: str = None
+    forecastStation: str = None
+
+    def __hash__(self):
+        return hash((self.city, self.state))
+
+API_KEY = 'YOUR API_KEY_HERE'
+CITY_LOCK = Lock()
+trackedCities = set()
+
+# First, let's create a table to manage the relevant cities to monitor
+columnNames = ["Timestamp", "State", "City", "Temp", "Humidity"]
+columnTypes = [dht.datetime, dht.string, dht.string, dht.float64, dht.float64]
+tableWriter = DynamicTableWriter(columnNames, columnTypes)
+
+CurrentData = tableWriter.getTable()
+
+# Bin the data by 30 minute and 1 hour intervals
+binnedData = CurrentData.updateView("bin30M=lowerBin(Timestamp, 30 * MINUTE)", "bin1Hr=lowerBin(Timestamp, 1 *HOUR)")
+
+# Compute some statistics
+binnedStats30 = binnedData.aggBy(as_list([ \
+    agg.AggMin("Min30Temp=Temp", "Min30Humid=Humidity"), \
+    agg.AggMax("Max30Temp=Temp", "Max30Humid=Humidity"), \
+    agg.AggAvg("Avg30Temp=Temp", "Avg30Humid=Humidity"), \
+    agg.AggFirst("bin1Hr")]), "State", "City", "bin30M")
+
+binnedStats60 = binnedData.aggBy(as_list([ \
+    agg.AggMin("Min60Temp=Temp", "Min60Humid=Humidity"), \
+    agg.AggMax("Max60Temp=Temp", "Max60Humid=Humidity"), \
+    agg.AggAvg("Avg60Temp=Temp", "Avg60Humid=Humidity")]), "State", "City", "bin1Hr")
+
+combinedStats = binnedStats30.naturalJoin(binnedStats60, "State,City,bin1Hr")
+
+# Now make a table containing the last values of all the relevant intervals for each city.
+LastByCityState = combinedStats.lastBy("State", "City") \
+    .dropColumns("bin30M", "bin1Hr") \
+    .naturalJoin(CurrentData.lastBy("State", "City"), "State,City") \
+    .moveColumnsUp("Timestamp", "State", "City", "Temp", "Humidity")
+
+
+# Borrowed from https://stackoverflow.com/questions/27928/calculate-distance-between-two-latitude-longitude-points-haversine-formula
+def distance(lat1, lon1, lat2, lon2):
+    p = pi/180
+    a = 0.5 - cos((lat2-lat1)*p)/2 + cos(lat1*p) * cos(lat2*p) * (1-cos((lon2-lon1)*p))/2
+    return 12742 * asin(sqrt(a)) #2*R*asin...
+
+# Locate the Lat/Lon of a particular city state
+def geoLocate(cityName) -> Location:
+    if len(cityName) <= 0:
+        raise ValueError("Place name must be present")
+
+    # First go figure out Lat/Lon so we can pass that to NOAA and locate a station
+    geoResp = requests.get("https://maps.googleapis.com/maps/api/geocode/json", params={'address': cityName, 'key': API_KEY})
+    if geoResp.status_code != 200:
```

Comment:
general http thing: any 200-299 is "success", you probably have a specific error code for "not a valid place" like 404, and might also have something like 503 meaning "hi, load killed me, try back later" etc.